File: CfxGenerator/CustomSignatures/StringCollectionAsRetvalSignature.vb
Public Class StringCollectionAsRetvalSignature
    Inherits Signature

    Public Sub New(parent As ISignatureOwner, sd As Parser.SignatureData, api As ApiTypeBuilder)
        MyBase.New(parent, sd, api)
    End Sub


    Public Overrides ReadOnly Property ManagedArguments As Argument()
        Get
            Return New Argument() {MyBase.ManagedArguments(0)}
        End Get
    End Property

    Public Overrides ReadOnly Property PublicReturnType As ApiType
        Get
            Return MyBase.ManagedArguments(1).ArgumentType
        End Get
    End Property


    Public Overrides Sub EmitPublicCall(b As CodeBuilder)
        b.AppendLine("{0} {1} = new {0}();", MyBase.ManagedArguments(1).ArgumentType.PublicSymbol, MyBase.ManagedArguments(1).VarName)
        MyBase.ManagedArguments(1).EmitPrePublicCallStatements(b)
        b.AppendLine(String.Format("CfxApi.{0}(NativePtr, {1});", Owner.CfxApiFunctionName, MyBase.ManagedArguments(1).PublicUnwrapExpression))
        MyBase.ManagedArguments(1).EmitPostPublicStatements(b)
        b.AppendLine("return {0};", MyBase.ManagedArguments(1).VarName)
    End Sub

End Class

File: CfxGenerator/CustomSignatures/StringCollectionAsRetvalSignature.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

public class StringCollectionAsRetvalSignature : Signature {

    public StringCollectionAsRetvalSignature(Signature s)
        : base(s) {
    }

    public override Parameter[] ManagedParameters {
        get { return new Parameter[] { base.ManagedParameters[0] }; }
    }

    public override ApiType PublicReturnType {
        get { return base.ManagedParameters[1].ParameterType; }
    }

    public override void EmitPublicCall(CodeBuilder b, string apiClassName, string apiFunctionName) {
        b.AppendLine("{0} {1} = new {0}();", base.ManagedParameters[1].ParameterType.PublicSymbol, base.ManagedParameters[1].VarName);
        base.ManagedParameters[1].EmitPrePublicCallStatements(b);
        b.AppendLine(string.Format("CfxApi.{2}.{0}(NativePtr, {1});", apiFunctionName, base.ManagedParameters[1].PublicCallArgument, apiClassName));
        base.ManagedParameters[1].EmitPostPublicStatements(b);
        b.AppendLine("return {0};", base.ManagedParameters[1].VarName);
    }

    protected override void EmitExecuteInTargetProcess(CodeBuilder b, string apiClassName, string apiFunctionName) {
        var collectionType = base.ManagedParameters[1].ParameterType.AsStringCollectionType.ClassName;
        b.AppendLine("__retval = new {0}();", base.ManagedParameters[1].ParameterType.PublicSymbol);
        b.AppendLine("var list = StringFunctions.Alloc{0}();", collectionType);
        b.AppendLine("CfxApi.{0}.{1}(@this, list);", apiClassName, apiFunctionName);
        b.AppendLine("StringFunctions.{0}CopyToManaged(list, __retval);", collectionType);
        b.AppendLine("StringFunctions.Free{0}(list);", collectionType);
    }

}